src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/ShardedHttpExecutionServiceCluster.scala

```diff
@@ -0,0 +1,110 @@
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success, Try}
+
+
+class ShardedHttpExecutionServiceCluster (members: Map[ExecutionServiceId,ExecutionServiceDAO], dataSource: SlickDataSource) extends ExecutionServiceCluster {
+
+  // make a copy of the members map as an array for easy reads
+  private val memberArray:Array[ClusterMember] = (members map {case (id, dao) => ClusterMember(id, dao)}).toArray
+
+
+
+
+  // ====================
+  // facade methods
+  // ====================
+
+  // only used in tests??? TODO: DA update tests to use the same thing runtime does, or remove this.
+  def submitWorkflow(wdl: String, inputs: String, options: Option[String], userInfo: UserInfo): Future[ExecutionServiceStatus] =
+    members.values.head.submitWorkflow(wdl, inputs, options, userInfo)
+
+  // by nature, this is only called for workflows that have not yet been submitted.
+  // therefore, we want to send the workflows to the cromwell instance chosen
+  // by the routing algorithm defined in this class.
+  def submitWorkflows(workflowRecs: Seq[WorkflowRecord], wdl: String, inputs: Seq[String], options: Option[String], userInfo: UserInfo): Future[(ExecutionServiceId, Seq[Either[ExecutionServiceStatus, ExecutionServiceFailure]])] = {
+    val targetMember = nextAvailableMember(workflowRecs)
+    targetMember.dao.submitWorkflows(wdl, inputs, options, userInfo) map {results =>
+      (targetMember.key, results)
+    }
+  }
+
+  // following are called on a workflow that has already been submitted.
+  // therefore, we want to use the cromwell instance that has been persisted
+  // onto that workflow.
+  def status(wfe: WorkflowExecution, userInfo: UserInfo): Future[ExecutionServiceStatus] =
+    getMember(wfe).flatMap {memb => memb.status(wfe.id, userInfo)}
+
+  def callLevelMetadata(wfe: WorkflowExecution, userInfo: UserInfo): Future[ExecutionMetadata] =
+    getMember(wfe).flatMap {memb => memb.callLevelMetadata(wfe.id, userInfo)}
+
+  def outputs(wfe: WorkflowExecution, userInfo: UserInfo): Future[ExecutionServiceOutputs] =
+    getMember(wfe).flatMap {memb => memb.outputs(wfe.id, userInfo)}
+
+  def logs(wfe: WorkflowExecution, userInfo: UserInfo): Future[ExecutionServiceLogs] =
+    getMember(wfe).flatMap {memb => memb.logs(wfe.id, userInfo)}
+
+  def abort(wfe: WorkflowExecution, userInfo: UserInfo): Future[Try[ExecutionServiceStatus]] =
+    getMember(wfe).flatMap {memb => memb.abort(wfe.id, userInfo)}
+
+
+  // ====================
+  // facade-to-cluster entry points
+  // ====================
+  // for an already-submitted workflow, get the instance to which it was submitted
+  private def getMember(wfe: WorkflowExecution):Future[ExecutionServiceDAO] = wfe.executionServiceId match {
+    case Some(execId) => Future(getMember(execId))
+    case None => {
+      dataSource.inTransaction { dataAccess =>
+        dataAccess.workflowQuery.getExecutionServiceKey(wfe.id) map {execIdOption =>
```

Comment:
why wouldn't we already have the key? I am concerned about nested transactions and a query per workflow model.